### PR TITLE
Top recoil hook cleaned

### PR DESCRIPTION
**Pull request by vslokenb**
#### PR description:

Integration of recoilToTop code into CMSSW Pythia8 interface. Allows us the option to treat the top quark as the recoiler during gluon emission in t -> bW decay. Expect to see differences between recoilToTop and the current recoilToColoured options, particularly in B hadron sensitive variables. PR includes header file containing actual function, as well as edits to existing files Pythia8Hadronizer.cc and Py8Interface.cc to incorporate said changes. Python config file to for easy test and run of recoilToTop is included.

#### PR validation:

Compiles properly using `scram b -j 8` and `scram b codeformat` Private samples generated with careful analysis conducted in Rivet to compare between the different recoiler options. No issues in using each of the options when running CMSSW. Only validated on CMSSW_13_2_0_pre1.

### CODE REVIEW

**Comment by Saptaparna** [comment]
please test

**Comment by Saptaparna** [comment]
please test

**Comment by perrotta** [comment]
@cms-sw/generators-l2 please evaluate which other backports may be needed, besides the already submitted #42202 for CMSSW_10_6_X

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        if (event[iEmt].col() != event[iTop].col())
+          return false;
+      } else {
+        if (event[iEmt].acol() != event[iTop].acol())
+          return false;
+      }
+
+      // Recoiler should now be a W, else something is wrong.
+      if (event[iRec].idAbs() != 24) {
+        cout << " ERROR: recoiler is " << event[iRec].id() << endl;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:98]: Use LogMessages, not `cout`, in the production code

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+      useOldDipole = useOldDipoleIn;
+      doList = doListIn;
+      // Constructor also creates some histograms for analysis inside User Hook.
+      wtCorr = new Hist("corrective weight", 100, 0., 2.);
+    }
+
+    // Destructor prints histogram.
+    ~TopRecoilHook() override {
+      if (doTopRecoil)
+        cout << *wtCorr;
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:38]: Either protect, remove, or use LgMessages instead

File: GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc [hunk 3]
```diff
@@ -19,6 +19,7 @@ using namespace Pythia8;
 
 #include "GeneratorInterface/Pythia8Interface/plugins/ReweightUserHooks.h"
 #include "GeneratorInterface/Pythia8Interface/interface/CustomHook.h"
+#include "GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h"
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/Pythia8Hadronizer.cc:22]:
```suggestion
#include "ReweightUserHooks.h"
#include "GeneratorInterface/Pythia8Interface/interface/CustomHook.h"
#include "TopRecoilHook.h"
```

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h [hunk 4]
```diff
@@ -0,0 +1,138 @@
+// TopRecoilHook.h is a part of the PYTHIA event generator.
```
Comment (perrotta) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:1]: Is this code modified with respect to the one in PYTHIA?

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h (hunk 2 above)
Comment (mseidel42) [reply GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:38]: Hi Andrea, yes, couts will be removed!

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h (hunk 4 above)
Comment (mseidel42) [reply GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:1]: It is not officially in Pythia yet, hopefully there will be a simple switch in the future.
But this is what the authors distributed to ATLAS as well.

File: GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h
Comment (menglu21) [inline GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:1]: Hi @mseidel42 , could you please remind me the initial thread of this hooker, sorry if we discussed this before e.g. in the GEN meeting

Comment (mseidel42) [reply GeneratorInterface/Pythia8Interface/plugins/TopRecoilHook.h:1]: Hi Meng, Laurids and Simone briefly discussed it last month: https://indico.cern.ch/event/1293969/contributions/5438463/attachments/2668515/4624955/GEN%20Meeting%20190623%20v2.pdf#page=17
(studies done in stand-alone Pythia)

**Comment by Saptaparna** [comment]
@mseidel42 As usual, trying to make sure that this user-hook will not alter anything else? Looks like it is pretty standalone. Let me know if this is accurate.

**Comment by mseidel42** [comment]
> @mseidel42 As usual, trying to make sure that this user-hook will not alter anything else? Looks like it is pretty standalone. Let me know if this is accurate.

Hi Sapta, yes, it's only active when `TopRecoilHook:doTopRecoilIn = on`.

**Comment by menglu21** [comment]
> @cms-sw/generators-l2 please evaluate which other backports may be needed, besides the already submitted #42202 for CMSSW_10_6_X

Hi @perrotta CMSSW_12_4_X should also need for Run3 production

**Comment by menglu21** [comment]
+1

**Comment by menglu21** [comment]
seems I should assign the test again, should I do "-1" and launch the test

**Comment by rappoccio** [comment]
+1

**Pull request merged**